sycl/include/sycl/nd_item.hpp

```diff
@@ -38,6 +39,273 @@ namespace ext::oneapi::experimental {
+                                        get_group_range(), Id);
+  }
+
+  sub_group get_sub_group() const { return sub_group(); }
+
+  size_t __SYCL_ALWAYS_INLINE get_group(int Dimension) const {
+    // Why Size? It returns group id!!!
+    size_t Size = get_group()[Dimension];
+    __SYCL_ASSUME_INT(Size);
+    return Size;
+  }
+
+  size_t __SYCL_ALWAYS_INLINE get_group_linear_id() const {
+    size_t Id = get_global_linear_id() / get_local_range().size();
+    __SYCL_ASSUME_INT(Id);
+    return Id;
+  }
+
+  range<Dimensions> get_group_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initNumWorkgroups<Dimensions, range<Dimensions>>();
+#else
+    return {};
+#endif
+  }
+
+  size_t __SYCL_ALWAYS_INLINE get_group_range(int Dimension) const {
+    size_t Range = get_group_range()[Dimension];
+    __SYCL_ASSUME_INT(Range);
+    return Range;
+  }
+
+  range<Dimensions> get_global_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initGlobalSize<Dimensions, range<Dimensions>>();
+#else
+    return {};
+#endif
+  }
+
+  size_t get_global_range(int Dimension) const {
+    size_t Val = get_global_range()[Dimension];
+    __SYCL_ASSUME_INT(Val);
+    return Val;
+  }
+
+  range<Dimensions> get_local_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initWorkgroupSize<Dimensions, range<Dimensions>>();
+#else
+    return {};
+#endif
+  }
+
+  size_t get_local_range(int Dimension) const {
+    size_t Id = get_local_range()[Dimension];
+    __SYCL_ASSUME_INT(Id);
+    return Id;
+  }
+
+  __SYCL2020_DEPRECATED("offsets are deprecated in SYCL 2020")
+  id<Dimensions> get_offset() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initGlobalOffset<Dimensions, id<Dimensions>>();
+#else
+    return {};
+#endif
+  }
+
+  nd_range<Dimensions> get_nd_range() const { return nd_range<Dimensions>(); }
+
+  void barrier(access::fence_space accessSpace =
+                   access::fence_space::global_and_local) const {
+    uint32_t flags = _V1::detail::getSPIRVMemorySemanticsMask(accessSpace);
+    __spirv_ControlBarrier(__spv::Scope::Workgroup, __spv::Scope::Workgroup,
+                           flags);
+  }
+
+  /// Executes a work-group mem-fence with memory ordering on the local address
+  /// space, global address space or both based on the value of \p accessSpace.
+  template <access::mode accessMode = access::mode::read_write>
+  __SYCL2020_DEPRECATED("use sycl::atomic_fence() free function instead")
+  void mem_fence(
+      typename std::enable_if_t<accessMode == access::mode::read ||
+                                    accessMode == access::mode::write ||
+                                    accessMode == access::mode::read_write,
+                                access::fence_space>
+          accessSpace = access::fence_space::global_and_local) const {
+    uint32_t flags = detail::getSPIRVMemorySemanticsMask(accessSpace);
```

Comment:
@maarquitos14, note, `mem_fence` is handle differently from other `group` methods.
I inlined the implementation, rather than making `group` class method static. It's because it's much simpler than `async_*` and `wait` methods.
On top of that, I pass `accessSpace` argument to SPIR-V built-in, whereas current `nd_item` implementation ignores it. The argument is ignored in `nd_item` method, but `TODO` comment below is from `group` class member implementation. It looks to me that the comment from `group` method tries to explain the code from `nd_item` method.
Are there any plans to clean things up? I'm not sure if I also should ignore `accessSpace` argument or not.

Reply:
Firstly, sorry for the delay, I have been off. 

Honestly, I'm not familiar with this issue, but my understanding of the comment is that it doesn't matter what we pass to SPIR-V because SPIR-V can't easily discern between load-only or store-only operations. 

 Let's see if other folks from @intel/llvm-reviewers-runtime can help here.